phi_four.py

```diff
@@ -0,0 +1,233 @@
+        output = model(phi)
+
+        model.zero_grad() # get rid of stored gradients
+        loss = shifted_kl(output, target)
+        loss.backward() # calc gradients
+
+        optimizer.step()
+        scheduler.step(loss)
+
+        if (i%50) == 0:
+            pbar.set_description(f"loss: {loss.item()}")
+
+def sample(model, a_left, b_right, n_large, target_length):
+    r"""
+    Sample using Metroplis-Hastings algorithm from a large number of phi configurations.
+    We calculate an A = min(1, \frac{\tilde p(phi^i)}{p(phi^i)} \frac{p(phi^j)}{\tilde p(phi^j)})
+    Where i is the index of the current phi, and j is the index of the update proposal phi.
+    We then generate a random number u, and if u <= A then we accept the update and i = j for the
+    next proposal, phi^i is added to the chain of samples, and a new j is picked. If the update
+    is rejected, then i=i for the next proposal and a new j is picked. We continue until the chain
+    has the desired length.
+    """
+
+    with torch.no_grad(): # don't want gradients being tracked in sampling stage
+        z = torch.randn((n_large, N_UNITS)) # random z configurations
+        phi = model.inverse_map(z) # map using trained model to phi
+        p_tilde = torch.exp(model.forward(phi)) # probabilities of generated phis using trained model
+        p = torch.exp(-target_distribution_s(phi, m_sq, l, shift)) # probabilities of phis from target pdf
+        chain_len = 0 # intialise current chain length
+        sample_distribution = torch.Tensor(target_length, N_UNITS) # intialise tensor to store samples
+        accepted, rejected = 0,0 # track accept/reject statistics
+        i = np.random.randint(n_large) # random index to start sampling from configurations tensor
+        used = [] # track which configurations have been added to the chain
+        used.append(i) 
+        while chain_len < target_length:
+            j = np.random.randint(n_large) # random initial phi^j for update proposal
+            while j in used: # make sure we don't pick a phi we have already used
+                j = np.random.randint(n_large)
+            ratio = (p_tilde[i]/p_tilde[j]) * (p[j]/p[i]) # calculate the ratio in A
+            A = min(1, ratio) # set A
+            u = np.random.uniform() # pick a random u for comparison
+            if u <= A:
+                sample_distribution[chain_len,:] = phi[i,:] # add to chain if accepted
+                chain_len += 1 
+                i = j # update i for next proposal
+                used.append(i)
+                accepted += 1
+            else:
+                rejected += 1
+    print('Accepted: '+str(accepted)+', Rejected:'+str(rejected))
+    print('Fraction accepted: '+str(accepted/(accepted+rejected)))
+    return sample_distribution
+
+def main():
+    # set seed, hopefully result is reproducible
+    torch.manual_seed(0)
+    a_left, b_right = split_transformation(L)
```

Comment:
I'm pretty sure we don't want to use these. I had them in the other training example because my action was defined for a flattened state, where as the model returns a flattened split state i.e (even sites, odd sites). Since your action is defined for split states then the interaction pattern is not nearest neighbour geometry but something slightly different.

I think since you use this transformation everywhere then it technically won't matter if everything is self consistent although we should still sort this out so it's obvious what we're doing